src/Lean/Elab/Tactic/Ext/Attr.lean

```diff
@@ -0,0 +1,111 @@
+/-
+Copyright (c) 2021 Gabriel Ebner. All rights reserved.
+Released under Apache 2.0 license as described in the file LICENSE.
+Authors: Gabriel Ebner, Mario Carneiro
+-/
+import Lean.Elab.Tactic.RCases
+import Lean.Meta.DiscrTree
+import Lean.Elab.Command
+
+namespace Lean.Elab.Tactic.Ext
+open Lean Meta
+
+/-- `declare_ext_theorems_for A` declares the extensionality theorems for the structure `A`. -/
+syntax "declare_ext_theorems_for " ("(" &"flat" " := " term ") ")? ident (ppSpace prio)? : command
+
+/-- Information about an extensionality theorem, stored in the environment extension. -/
+structure ExtTheorem where
+  /-- Declaration name of the extensionality theorem. -/
+  declName : Name
+  /-- Priority of the extensionality theorem. -/
+  priority : Nat
+  /-- Key in the discrimination tree. -/
```

Comment:
Perhaps a bit should be said about what this key actually contains? It's the type for which the lemma applies, right?